lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs

```diff
@@ -537,42 +460,117 @@ getLegacyTransaction tid = do
+        withHeader FragmentTransaction $ do
+            putTransaction witTag
+
+    -- STAKE-DELEGATION = DLG-CERT TRANSACTION STAKE-AUTH
+    MkFragmentStakeDelegation witTag dlgTag account credentials ->
+        withHeader FragmentDelegation $ do
+            putDelegationCertificate account dlgTag
+            putTransaction witTag
+            putStakeAuthentication account credentials
   where
-    putInput (TxIn inputId inputIx, coin) = do
-        -- NOTE: special value 0xff indicates account spending
-        -- only old utxo/address scheme supported for now
+    -- TRANSACTION =
+    --     SIZE-ELEMENT-8BIT ; number of inputs
+    --     SIZE-ELEMENT-8BIT ; number of outputs
+    --     *INPUT            ; as many as indicated in the number of inputs
+    --     *OUTPUT           ; sa many as indicated in the number of outputs
+    --     *WITNESS          ; as many as indicated in the number of inputs
+    putTransaction witTag = do
+        putInputsOutputs
+        mapM_ (`putWitness` witTag) (snd <$> inputs)
+
+    putInputsOutputs = do
+        guardLength (<= maxNumberOfInputs) inputs
+        guardLength (<= maxNumberOfOutputs) outputs
+        putWord8 $ toEnum $ length inputs
+        putWord8 $ toEnum $ length outputs
+        mapM_ putInput (fst <$> inputs)
+        mapM_ putOutput outputs
+
+    -- INPUT             = INPUT-UTXO / INPUT_ACCOUNT
+    -- INPUT-UTXO        = IDX VALUE FRAGMENT-ID
+    -- INPUT-ACCOUNT     = %xff VALUE UNTAG-ACCOUNT-ID
+    -- IDX               = %x00-fe
+    putInput (TxIn inputId inputIx, inputValue) = do
+        -- TODO invariant about inputIx /= 0xff
         putWord8 . toEnum . fromEnum $ inputIx
-        putWord64be $ getCoin coin
+        putWord64be $ getCoin inputValue
         putByteString $ getHash inputId
 
-    putOutput (TxOut address coin) = do
-        putAddress address
-        putWord64be $ getCoin coin
+    -- OUTPUT            = ADDRESS VALUE
+    putOutput (TxOut outputAddr outputValue) = do
+        putAddress outputAddr
+        putWord64be $ getCoin outputValue
+
+    -- WITNESS           = WITNESS-OLDUTXO / WITNESS-UTXO / WITNESS-ACCOUNT / WITNESS-MULTISIG
+    -- WITNESS-OLDUTXO   = %x00 LEGACY-XPUB LEGACY-SIGNATURE
+    -- WITNESS-UTXO      = %x01 ED25519-SIGNATURE
+    -- WITNESS-ACCOUNT   = %x02 SINGLE-ACNT-SIG
+    -- WITNESS-MULTISIG  = %x03 MULTI-ACNT-SIG
+    putWitness (xprv, Passphrase pwd) = \case
+        TxWitnessUTxO -> do
+            putByteString $ unXSignature $ sign pwd xprv msg
+        TxWitnessLegacyUTxO -> do
+            putByteString $ unXSignature $ sign pwd xprv msg
+            putByteString $ unXPub $ toXPub xprv
+        TxWitnessAccount ->
+            error "putWitness: TxWitnessAccount: not implemented"
+        TxWitnessMultisig ->
+            error "putWitness: TxWitnessMultisig: not implemented"
+      where
+        msg = mconcat
+            [ block0Hash
+            , blake2b256 $ BL.toStrict $ runPut putInputsOutputs
```

Comment:
As I've said already, I'm pretty sure this should be `blake2b256 (payload <> putInputsOutputs)`

Reply:
We currently have: 

```hs
            let payload = block0H <> getHash (signData inps outs)
```

Where

```hs
signData
    :: [(TxIn, Coin)]
    -> [TxOut]
    -> Hash "SignData"
signData inps outs =
    Hash $ blake2b256 $ BL.toStrict $ runPut $ do
        putTx inps outs
```

So I think this is correct, and clearer at the same time.

Reply:
I meant the payload as in where the certificate goes.